mir_connector/inorbit_mir_connector/src/connector.py

```diff
@@ -400,7 +398,139 @@ async def _execution_loop(self):
+                try:
+                    if isinstance(value, (int, float)):
+                        return float(value)
+                    s = str(value)
+                    # Extract first decimal/float-like token
+                    import re
+                    m = re.search(r"-?\d+(?:[.,]\d+)?", s)
+                    if not m:
+                        return None
+                    num = m.group(0).replace(",", ".")
+                    return float(num)
+                except (ValueError, AttributeError):
+                    # Expected parsing errors for malformed numbers
+                    return None
+
+            def _to_gb(val: object, key: str) -> float | None:
+                n = _parse_number(val)
+                if n is None:
+                    return None
+                k = key.lower()
+                if "[gb]" in k:
+                    return n
+                if "[mb]" in k:
+                    return n / 1024.0
+                if "[b]" in k:
+                    return n / (1024.0 * 1024.0 * 1024.0)
+                # Assume already in GB if unit not specified
+                return n
+            # Battery from diagnostics (/Power System/Battery), fallback to status
+            batt_vals = (diagnostics.get('/Power System/Battery', {}) or {}).get('values', {})
+            remaining_pct = None
+            remaining_sec = None
+            for k, v in batt_vals.items():
+                if 'Remaining battery capacity' in k:
+                    remaining_pct = float(v)
+                elif 'Remaining battery time [sec]' in k:
+                    remaining_sec = float(v)
+            if remaining_pct is not None:
+                key_values['battery percent'] = _to_inorbit_percent(remaining_pct)
+            if remaining_sec is not None:
+                key_values['battery_time_remaining'] = int(remaining_sec)
+
+            # CPU load from diagnostics
+            cpu_vals = (diagnostics.get('/Computer/PC/CPU Load', {}) or {}).get('values', {})
+            avg_cpu = None
+            for k, v in cpu_vals.items():
+                if 'Average CPU load' in k and '30 second' not in k and '3 minut' not in k:
+                    avg_cpu = float(v)
+                    break
+            if avg_cpu is not None:
+                key_values['cpu_usage_percent'] = _to_inorbit_percent(avg_cpu)
+
+            # CPU temperature from diagnostics
+            cpu_temp_vals = (diagnostics.get('/Computer/PC/CPU Temperature', {}) or {}).get('values', {})
+            pkg_temp = None
+            for k, v in cpu_temp_vals.items():
+                if 'Package id' in k:
+                    pkg_temp = float(v)
+                    break
+            if pkg_temp is not None:
+                key_values['temperature_celsius'] = pkg_temp
+
+            # Helper function to calculate usage percentage from diagnostics
+            def _calculate_usage_percent(diagnostic_path: str, key_name: str) -> None:
```

Comment:
I'd move all of these helper functions outside of this loop (on a separate file, even better). They don't use `self` and having them here makes unittesting harder